File: 프로그래머스/2/92342. 양궁대회/양궁대회.py
def solution(n, info):
    from copy import deepcopy

    max_diff = 0
    best = [-1]

    def backtrack(idx, arrows_left, lion):
        nonlocal max_diff, best
        if idx == 11:
            if arrows_left > 0:
                lion[10] += arrows_left  # 남은 화살을 0점에 할당
            # 점수 계산
            lion_score = 0
            peach_score = 0
            for i in range(11):
                if info[i] == 0 and lion[i] == 0:
                    continue
                if lion[i] > info[i]:
                    lion_score += 10 - i
                else:
                    peach_score += 10 - i
            diff = lion_score - peach_score
            if diff > 0:
                if diff > max_diff:
                    max_diff = diff
                    best = deepcopy(lion)
                elif diff == max_diff:
                    # 더 낮은 점수를 많이 맞힌 경우 선택
                    for i in range(10, -1, -1):
                        if lion[i] > best[i]:
                            best = deepcopy(lion)
                            break
                        elif lion[i] < best[i]:
                            break
            if arrows_left > 0:
                lion[10] -= arrows_left  # 초기 상태로 복구
            return

        # 옵션 1: 라이언이 현재 점수를 이기기 위해 화살을 할당
        if arrows_left > info[idx]:
            lion[idx] = info[idx] + 1
            backtrack(idx + 1, arrows_left - (info[idx] + 1), lion)
            lion[idx] = 0  # 초기 상태로 복구

        # 옵션 2: 라이언이 현재 점수를 이기지 않음
        backtrack(idx + 1, arrows_left, lion)

    initial_lion = [0] * 11
    backtrack(0, n, initial_lion)

    return best

File: 프로그래머스/2/92342. 양궁대회/양궁대회.js
function solution(n, info) {
  const CASES = 3 ** 11;
  let maxDiff = 0;
  let result = [-1];
  let PEACHSCORE = 0;
  for (let i = 0; i < 11; i++) {
    if (info[i] !== 0) {
      PEACHSCORE += 10 - i;
    }
  }
  for (let i = 0; i < CASES; i++) {
    let brute = i;
    let shot = 0;
    let lionScore = 0,
      peachScore = PEACHSCORE;
    const archer = new Array(11).fill(0);
    for (let j = 0; j < 11; j++) {
      const type = brute % 3;
      const peachShot = info[j];
      // 0: 무승부, 1: 승, 2: 패
      switch (type) {
        case 0:
          shot += peachShot;
          archer[j] = peachShot;
          break;
        case 1:
          shot += peachShot + 1;
          archer[j] = peachShot + 1;
          lionScore += 10 - j;
          if (peachShot !== 0) {
            peachScore -= 10 - j;
          }
          break;
      }
      brute = Math.floor(brute / 3);
    }
    if (shot < n) {
      archer[10] += n - shot;
      shot = n;
    }
    const diff = lionScore - peachScore;
    if (shot === n && diff > 0 && diff >= maxDiff) {
      if (maxDiff === diff) {
        let canReplace = true;
        for (let j = 10; j >= 0; j--) {
          if (archer[j] > result[j]) {
            break;
          }
          if (archer[j] < result[j]) {
            canReplace = false;
            break;
          }
        }
        if (!canReplace) {
          continue;
        }
      }
      maxDiff = diff;
      result = archer;
    }
  }
  return result;
}
